### PR TITLE
ENH Add ability for admin to revoke all sessions

### PR DESCRIPTION
**Pull request by GuySartorelli**
Note that the `onBeforeRemoveLoginSession` extension hook (see `LoginSessionController::remove()`) is not called. The thinking is that this task is effectively for emergency situations anyway to prevent an identified actively logged in potential attacker from making additional changes - the aim is to outright remove all current sessions.
This is also consistent with the `GarbageCollectionTask`. 

## Acceptance criteria (from parent issue, for the sake of transparency)
- Create a new task on session manager that invalidates all sessions.
- The task can only be run run by admin
  - In this case this is handled by the regular permission check that prevents non-admin from calling dev tasks on production
- Task invalidates all sessions, including the one of the user running it.
- Validates what happens if you destroy a thousand session in one go.
  - Checked by creating a temporary task that created a thousand sessions, then running the new task in the browser. Worked fine. Not sure that a unit test is required for this but I can see about adding one if someone disagrees.
- Validates that remember me token don't allow you to login anymore.
  - The actual tests for this is `SilverStripe\Security\Tests\RememberLoginHashTest::testClear()` in framework - for the purposes of this PR, testing that they're removed is sufficient.
- Validates that this work with hybrid sessions as well.
  - Checked manually in browser and by running the tests with hybrid sessions installed. I'm not sure where we'd add unit tests for this if we were to add them.

## Parent issue:
- silverstripeltd/product-issues#525

### CODE REVIEW

File: src/Tasks/InvalidateAllSessionsTask.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+use SilverStripe\Dev\BuildTask;
+use SilverStripe\Security\RememberLoginHash;
+use SilverStripe\SessionManager\Models\LoginSession;
+
+class InvalidateAllSessionsTask extends BuildTask
+{
+    /**
+     * @var string
+     */
+    private static $segment = 'InvalidateAllSessions';
```
Comment (emteknetnz) [inline src/Tasks/InvalidateAllSessionsTask.php]: Get rid of the doc blocks and use typed properties - https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.typed-properties.  We use 7.4 as minimum now

File: src/Tasks/InvalidateAllSessionsTask.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * @var string
+     */
+    protected $description = 'Removes all login sessions and “remember me” hashes (including yours) from the database';
+
+    /**
+     * @param HTTPRequest $request
+     */
+    public function run($request)
```
Comment (emteknetnz) [inline src/Tasks/InvalidateAllSessionsTask.php:27]:
```suggestion
    public function run(HTTPRequest $request)
```

File: src/Tasks/InvalidateAllSessionsTask.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace SilverStripe\SessionManager\Tasks;
+
+use SilverStripe\Dev\BuildTask;
```
Comment (emteknetnz) [inline src/Tasks/InvalidateAllSessionsTask.php:6]:
```suggestion
use SilverStripe\Control\HTTPRequest;
use SilverStripe\Dev\BuildTask;
```

File: src/Tasks/InvalidateAllSessionsTask.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * @var string
+     */
+    protected $title = 'Invalidate All Login Sessions Task';
+
+    /**
+     * @var string
+     */
+    protected $description = 'Removes all login sessions and “remember me” hashes (including yours) from the database';
```
Comment (emteknetnz) [inline src/Tasks/InvalidateAllSessionsTask.php]:
```suggestion
    protected $description = 'Removes all login sessions and "remember me" hashes (including yours) from the database';
```
Looks like your IDE was being too cute with pretty quotes

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+        // Log in as member 1 and make them active
+        $request = $this->buildRequestMock('dev/tasks/' . InvalidateAllSessionsTask::config()->get('segment'));
+        $request->method('getIP')->willReturn('192.168.0.1');
+        $member = $this->objFromFixture(Member::class, 'member1');
+        $handler = Injector::inst()->get(AuthenticationHandler::class);
+        $handler->login($member, true, $request);
+
+        // Ensure fixtured sessions are created and at least one login hash is generated
+        $this->assertNotSame(0, LoginSession::get()->count(), 'There are no LogginSessions');
```
Comment (emteknetnz) [inline tests/php/Tasks/InvalidateAllSessionsTaskTest.php]:
```suggestion
        $this->assertNotSame(0, LoginSession::get()->count(), 'There are no LoginSessions');
```

Comment (emteknetnz) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php]: Seems like we should be able to `assertSame(1, ` instead of `assertNotSame(0, ` ?

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        $middleware->process(
+            $request,
+            function () use (&$next) {
+                $next = true;
+            }
+        );
+        // Assert user who initiated the request's session was also invalidated
+        $this->assertNull(
+            Security::getCurrentUser(),
+            'user was logged out because their session was invalidated'
```
Comment (emteknetnz) [inline tests/php/Tasks/InvalidateAllSessionsTaskTest.php]:
```suggestion
            'User was not logged out when their session was invalidated'
```

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+SilverStripe\Security\Member:
+  member1:
+    FirstName: 'Andre'
+    Email: 'andre@example.org'
+  member2:
+    FirstName: 'Garion'
+    Email: 'garion@example.org'
```
Comment (emteknetnz) [inline tests/php/Tasks/InvalidateAllSessionsTaskTest.yml]:
```suggestion
```
Not using this one

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        $this->assertSame(0, LoginSession::get()->count(), 'There is at least one LoginSession');
+        $this->assertSame(0, RememberLoginHash::get()->count(), 'There is at least one RememberLoginHash');
+
+        // Navigate to re-validate current session
+        $middleware = new LoginSessionMiddleware(new Url('no-match'));
+        $next = false;
+        $middleware->process(
+            $request,
+            function () use (&$next) {
+                $next = true;
```
Comment (emteknetnz) [inline tests/php/Tasks/InvalidateAllSessionsTaskTest.php]:
```suggestion
                // no-op
```
We don't need the $next var?

**Review by emteknetnz — CHANGES_REQUESTED** [review]

**Comment by emteknetnz** [comment]
Re the hybridsessions AC
> Checked manually in browser and by running the tests with hybrid sessions installed

So, what was validated here?  Did you confirm that session data was removed the cookies?  Hybrid sessions is a strange one, it's easiest to test by commenting out [this line](https://github.com/silverstripe/silverstripe-hybridsessions/blob/2/_config/sessions.yml#L8) to ensure the database is used.  Maybe try that, ensure there's session data in the HybridSessionDataObject table, then run this task in the PR

If it's not deleted, something along the lines of a class_exists() check to see if CookieStore/DatabaseStore exist, then `singleton(DatabaseStore::class)->gc()` would work, though it's messy.

There should be something within the session-manager to "logout" a member that cascades down to the hybridsession stores to remove data

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php (hunk 8 above)
Comment (ChloeSartorelli) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php]: Fair enough. I just copied this from a similar test elsewhere in the module, and I don't understand middleware deeply enough yet to fully know how `$next` is used. 😅

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php (hunk 6 above)
Comment (GuySartorelli) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php:67]: Ahh, I must have misunderstood the messages - I thought it was intended to say "this is what went wrong" but it seems that it instead is meant to be "this is what we are testing"

Comment (ChloeSartorelli) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php]: The way tests in this module are written, the comments are "this is what this assertion is testing" as opposed to "this is what went wrong". I was trying to be consistent with the other tests in this module... should I update the other tests in this module to also have a message indicating what went wrong, as opposed to what is being tested?

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php (hunk 5 above)
Comment (ChloeSartorelli) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php]: I just care that it's not 0, not that it's explicitly 1. The test shouldn't fail if there are more sessions for whatever reason, because this assertion is just here to ensure there is something there which the task will soon clear out.

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php (hunk 6 above)
Comment (emteknetnz) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php]: er, the assertions directly above, the message are about saying what went wrong

`$this->assertNotSame(0, LoginSession::get()->count(), 'There are no LoginSessions');` aka
`$this->assertSame(1, LoginSession::get()->count(), 'There are no LoginSessions');`

File: src/Tasks/InvalidateAllSessionsTask.php (hunk 3 above)
Comment (GuySartorelli) [reply src/Tasks/InvalidateAllSessionsTask.php:5]: > Method 'SilverStripe\SessionManager\Tasks\InvalidateAllSessionsTask::run()' is not compatible with method 'SilverStripe\Dev\BuildTask::run()'

We can't add typehints here until they're also in framework.

File: src/Tasks/InvalidateAllSessionsTask.php (hunk 2 above)
Comment (ChloeSartorelli) [reply src/Tasks/InvalidateAllSessionsTask.php:27]: > Method 'SilverStripe\SessionManager\Tasks\InvalidateAllSessionsTask::run()' is not compatible with method 'SilverStripe\Dev\BuildTask::run()'

We can't add typehints here until they're also in framework.

File: src/Tasks/InvalidateAllSessionsTask.php (hunk 1 above)
Comment (ChloeSartorelli) [reply src/Tasks/InvalidateAllSessionsTask.php]: Wasn't sure we could do that with config - but I guess ultimately if someone puts a non-string in the config it'll just give them an error, which it should. So yup, good point.

File: tests/php/Tasks/InvalidateAllSessionsTaskTest.php (hunk 6 above)
Comment (ChloeSartorelli) [reply tests/php/Tasks/InvalidateAllSessionsTaskTest.php]: Yup, good point. I'm getting my wires crossed here.
I had a look and phpunit don't opine on what the message should be used for, so I've amended those to be consistent with this one, which in turn is consistent with the messages in other tests which are already in this module.

**Comment by GuySartorelli** [comment]
> There should be something within the session-manager to "logout" a member that cascades down to the hybridsession stores to remove data

This is probably what that `onBeforeRemoveLoginSession` hook is for (used in [`LoginSessionController::remove()`](https://github.com/silverstripe/silverstripe-session-manager/blob/d97e2bb9770dd699ab0fbe3ff47a2cee341ec49f/src/Controllers/LoginSessionController.php#L78) and discussed briefly in the description of this PR) so it looks like I will need to call that hook from this task after all.
I think instead of this module doing stuff to hybridsession, instead I will add a PR in hybridsession to use that extension hook if session manager is installed.

**EDIT**: No, we don't need to do this - see [my comment below](#issuecomment-1111728047).

File: src/Tasks/InvalidateAllSessionsTask.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+class InvalidateAllSessionsTask extends BuildTask
+{
+
+    private static string $segment = 'InvalidateAllSessions';
+
+    /**
+     * @var string
+     */
+    protected $title = 'Invalidate All Login Sessions Task';
```
Comment (ChloeSartorelli) [inline src/Tasks/InvalidateAllSessionsTask.php:17]: Can't strongly type this or `$description`
> Fatal error: Type of SilverStripe\SessionManager\Tasks\InvalidateAllSessionsTask::$title must not be defined (as in class SilverStripe\Dev\BuildTask)

**Comment by GuySartorelli** [comment]
I've just realised that hybridsession doesn't play into it at all. Hybridsession stores session data, but it doesn't touch authentication at all. That's handled by framework if session-manager isn't installed, and by session-manager if it is installed.

Imagine the scenario where a person has logged in and is accessing the CMS, but then their session gets revoked by the task in this PR. What happens when they refresh or navigate to a new page is this:
- A request comes through
- hybridsessions calls [session-set-save-handler](https://www.php.net/manual/en/function.session-set-save-handler.php) to set itself up to handle storage of session data
- hybridsessions middleware initialises the session - this allows fetching data stored for the current session such as form messages, etc. At this stage we're still just dealing with raw session data - no authentication has occurred.
- The [`AuthenticationMiddleware`](https://api.silverstripe.org/4/SilverStripe/Security/AuthenticationMiddleware.html) in framework processed, which results in `authenticateRequest` to be called on classes that implement [`AuthenticationHandler`](https://api.silverstripe.org/4/SilverStripe/Security/AuthenticationHandler.html)
- [`SessionAuthenticationHandler`](https://api.silverstripe.org/4/SilverStripe/Security/MemberAuthenticator/SessionAuthenticationHandler.html) sees that, based on the information stored for the current session, a user appears to be logged in.
- The `Member` from the session data set as the 'current' user via `Security::setCurrentUser()`.
- [`LoginSessionMiddleware`](https://github.com/silverstripe/silverstripe-session-manager/blob/1/src/Middleware/LoginSessionMiddleware.php) is processed. This gets the `Member` via `Security::getCurrentUser()` (if there was no member it would just return out at this point)
- `LoginSessionMiddleware` checks if there is a `LoginSession` object in the database that corresponds to the current session's ID. Because there isn't one, it logs the user out (which also revokes that session).
- Eventually hybridsessions garbage collection is run, which removes expired cookies and `HybridSessionDataObject`s.

In other words, we don't need to do anything special here to handle hybridsessions.
I'm still unsure whether we need to call the `onBeforeRemoveLoginSession` hook in the new task - I'm not sure what it's for, to be honest.
On the one hand calling it would potentially be a massive hit on performance for any sites with lots of sessions running, since they'd _all_ need to be fetched _and instantiated_ to call the hook.
On the other hand, is there a potential that some custom code relies on that hook to invalidate some weird custom session handling nonsense? I'm not sure.

**Comment by emteknetnz** [comment]
There's a PHPCS linting thing to fix in travis

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Could you also add a quick section in the readme saying this task is available.

**Comment by GuySartorelli** [comment]
Fixed the phpcs issue and added a blurb to the readme. Also added one about users revoking their own sessions, which didn't seem to be documented there.

**Review by emteknetnz — APPROVED** [review]
